infer/src/pulse/PulseFormula.ml

```diff
@@ -3849,15 +3947,22 @@ module Formula = struct
           (false, phi_new_eqs)
 
 
-    and and_normalized_atoms phi_new_eqs atoms =
-      SatUnsat.list_fold atoms ~init:(false, phi_new_eqs)
-        ~f:(fun (linear_changed, phi_new_eqs) atom ->
-          let+ changed', phi_new_eqs = and_normalized_atom phi_new_eqs atom in
+    and and_normalized_atoms (phi, new_eqs) atoms ~orig_atom ~add_term =
+      (* Possibly useful: test if variables in orig_atom are present in term_conds:
+         if so, add orig_atom in term_conds even if add_term is false.
+         Note: this may not be necessary thanks to Pulse fresh name creation which
+         is sufficient to witness the existence of a state update without tracking
+         intermediate constraints *)
+      let upd_phi = if add_term then and_termcond_atoms phi orig_atom else phi in
```

Comment:
If we could call `and_termcond_atoms` in client code instead of passing `orig_atom` and `add_term` all the way to `and_normalized_atoms` here the change would be much less invasive (and avoid the dubious adding of atoms during propagation above).